### PR TITLE
Add new bootstrap type: Equidistant point

### LINKED ISSUE #55: Give BootstrapAtData and BootstrapEquidistant a common parent
Some methods of BootstrapAtData and BootstrapEquidistant are equivalent. A common Bootstrap parent class can be created with these common methods. This limits the duplicated code between BootstrapAtData and BootstrapEquidistant.

### PR DESCRIPTION
**Pull request by PimMeulensteen**
Next to the already implemented training-data bootstrap method, this class uses equidistant points in stead. This is more in line with the code that was used for Scratch in Matlab.

### CODE REVIEW

File: lir/lrsystems/bootstraps.py [hunk 1]
```diff
@@ -97,3 +97,118 @@ def fit_transform(self, instances: FeatureData) -> LLRData:
...
+    def fit(self, instances: FeatureData) -> Self:
+        """Fit the bootstrap system to the provided instances.
+
+        param instances: FeatureData: The feature data to fit the bootstrap system on.
+        return Self: The fitted bootstrap system.
+        """
+
+        all_vals = []
+        if self.seed is not None:
+            np.random.seed(self.seed)
```
Comment (wowtor) [inline lir/lrsystems/bootstraps.py]: Deze is global en dat kan soms problemen geven.

Beter lokaal houden:
```py
rng = np.random.default_rng(self.seed)
```

File: lir/algorithms/bootstraps.py [hunk 2]
```diff
@@ -97,3 +97,118 @@ def fit_transform(self, instances: FeatureData) -> LLRData:
...
+
+        param instances: FeatureData: The feature data to fit the bootstrap system on.
+        return Self: The fitted bootstrap system.
+        """
+
+        all_vals = []
+        if self.seed is not None:
+            np.random.seed(self.seed)
+
+        feat_vals = instances.features.reshape(-1)
```
Comment (wowtor) [inline lir/algorithms/bootstraps.py]: Hiermee krijg je wel de juiste dimensies, maar als je meerdere kolommen aanbiedt klopt het niet meer met het aantal labels omdat je vector te lang is. Misschien eerst nog even een check doen op de dimensies met een mooie raise als het niet klopt.

File: lir/lrsystems/bootstraps.py [hunk 3]
```diff
@@ -97,3 +97,118 @@ def fit_transform(self, instances: FeatureData) -> LLRData:
...
+            self.n_points = len(instances)
+        equidistant_indices = np.linspace(
+            np.min(feat_vals), np.max(feat_vals), num=self.n_points, dtype=instances.features.dtype
+        ).reshape(-1, 1)
+
+        # Create FeatureData so the points can be used in the pipeline.
+        fd = FeatureData(features=equidistant_indices)
+
+        for _ in range(self.n_bootstraps):
+            sample_index = np.random.choice(len(instances), size=len(instances))
```
Comment (wowtor) [inline lir/lrsystems/bootstraps.py]:
```py
rng.choice(...)
```

File: lir/algorithms/bootstraps.py [hunk 4]
```diff
@@ -97,3 +97,118 @@ def fit_transform(self, instances: FeatureData) -> LLRData:
...
+
+        for _ in range(self.n_bootstraps):
+            sample_index = np.random.choice(len(instances), size=len(instances))
+            samples = instances[sample_index]
+            super().fit(samples)
+            all_vals.append(super().transform(fd).features.reshape(-1))
+
+        all_vals = np.stack(all_vals, axis=1)
+
+        intervals = np.quantile(all_vals, self.interval, axis=1)
```
Comment (wowtor) [inline lir/algorithms/bootstraps.py:122]: Eerder hadden we bedacht dat we de interval gingen zetten op de afwijking ten opzichte van best estimate. Anders ga je een beejte vreemd extrapoleren.
Dus 
```py
lower = intervals[0] - best_on_equidistant
upper = intervals[1] - best_on_equidistant
```

File: lir/algorithms/bootstraps.py [hunk 5]
```diff
@@ -97,3 +97,118 @@ def fit_transform(self, instances: FeatureData) -> LLRData:
...
+        self.f_interval_lower = interp1d(
+            best_on_equidistant, lower, bounds_error=False, fill_value=(lower[0], lower[-1])
+        )
+        self.f_interval_upper = interp1d(
+            best_on_equidistant, upper, bounds_error=False, fill_value=(upper[0], upper[-1])
+        )
+
+        return self
+
+    def transform(self, instances: FeatureData) -> LLRData:
```
Comment (wowtor) [inline lir/algorithms/bootstraps.py]: Deze code in `transform()` en `fit_transform()` is (denk ik?) hetzelfde als in de andere boostrapmethode. Misschien verplaatsen naar gedeelde parent?

File: lir/algorithms/bootstraps.py [hunk 6]
```diff
@@ -97,3 +97,118 @@ def fit_transform(self, instances: FeatureData) -> LLRData:
         return LLRData: The transformed feature data with best estimate and confidence intervals.
         """
         return self.fit(instances).transform(instances)
+
+
+class BootstrapEquidistant(Pipeline):
+    """Bootstrap system that estimates confidence intervals around the best estimate of a pipeline.
+    This bootstrap system creates bootstrap samples from the training data, fits the pipeline on each sample,
+    and then computes confidence intervals for the pipeline outputs based on the variability across the bootstrap
+    samples.
```
Comment (wowtor) [inline lir/algorithms/bootstraps.py]: Uit deze beschrijving blijkt niet heel duidelijk wat het verschil is met de andere bootstrapmethode. Daarnaast is het denk ik goed om de beperking van deze aan te geven, naamelijk dat het alleen werkt op 1-dimensionale data.

**Review by wowtor — CHANGES_REQUESTED** [review]

File: lir/algorithms/bootstraps.py [hunk 7]
```diff
@@ -97,3 +97,118 @@ def fit_transform(self, instances: FeatureData) -> LLRData:
         return LLRData: The transformed feature data with best estimate and confidence intervals.
         """
         return self.fit(instances).transform(instances)
+
+
+class BootstrapEquidistant(Pipeline):
```
Comment (wowtor) [inline lir/algorithms/bootstraps.py]: Deze staat onder "lrsystems" maar eigenlijk is dat misschien niet een logische plek. Beter "transform" of "algorithms".

File: lir/algorithms/bootstraps.py (hunk 5 above)
Comment (PimMeulensteen) [reply lir/algorithms/bootstraps.py]: Dat lijkt me goed! Ik ga dat wel in een latere PR doen om het een beetje gescheiden te houden.

Comment (PimMeulensteen) [reply lir/algorithms/bootstraps.py]: Zie #55

File: lir/algorithms/bootstraps.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+        param instances: FeatureData: The feature data to fit the bootstrap system on.
+        return Self: The fitted bootstrap system.
+        """
+
+        all_vals = []
+        rng = np.random.default_rng(self.seed)
+
+        # Raise an exception if the dimensions are not what we expect.
+        if instances.features.ndim != 2:
```
Comment (wowtor) [inline lir/algorithms/bootstraps.py]: Deze test doet maar de helft, want wat je wil is dat hij 1 kolom heeft.

```suggestion
        if instances.features.ndim != 2 or instances.features.shape[1] != 1:
            raise ValueError(f'expected 2D feature array with 1 column; found shape: {instances.features.shape}')
```

File: lir/algorithms/bootstraps.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        """
+
+        all_vals = []
+        rng = np.random.default_rng(self.seed)
+
+        for _ in range(self.n_bootstraps):
+            sample_index = rng.choice(len(instances), size=len(instances))
+            samples = instances[sample_index]
+            super().fit(samples)
+            all_vals.append(super().transform(instances).features.reshape(-1))
```
Comment (wowtor) [inline lir/algorithms/bootstraps.py]: Dit is de enige regel die verschillend is tussen beide klassen. Gedeelde base class met abstract method `def get_bootstrap_points(instances: FeatureData) -> FeatureData` (of iets van die strekking), en je kan de rest van de code delen. Weet je zeker dat je dit in een aparte PR wil doen?

File: tests/test_bootstrap.py [hunk 10]
```diff
@@ -14,14 +14,18 @@ def test_traindata_bootstrap():
...
+    results_equidistant = BootstrapEquidistant(steps).fit(feature_data).transform(feature_data)
 
     # Check that the llr values are within the inteval it has calculated.
-    assert np.all(results.llrs > results.llr_intervals[:, 0])
-    assert np.all(results.llrs < results.llr_intervals[:, 1])
+    assert np.all(results_data.llrs > results_data.llr_intervals[:, 0])
+    assert np.all(results_data.llrs < results_data.llr_intervals[:, 1])
+
+    assert np.all(results_equidistant.llrs > results_equidistant.llr_intervals[:, 0])
+    assert np.all(results_equidistant.llrs < results_equidistant.llr_intervals[:, 1])
```
Comment (wowtor) [inline tests/test_bootstrap.py]: nog een testje toevoegen voor de extrapolatie?

**Review by wowtor — CHANGES_REQUESTED** [review]

File: lir/algorithms/bootstraps.py (hunk 9 above)
Comment (PimMeulensteen) [reply lir/algorithms/bootstraps.py]: Het liefst wel, denk ik. Om twee redenen: de doorvoertijd van de PR laag houden, en ik verwacht dat de twee PR's los makkelijker te begrijpen zijn dan samen.

Comment (wowtor) [reply lir/algorithms/bootstraps.py]: Dat kan je ook bereiken in een losse commit binnen de PR

File: tests/test_bootstrap.py (hunk 10 above)
Comment (PimMeulensteen) [reply tests/test_bootstrap.py]: Yes!

File: lir/algorithms/bootstraps.py (hunk 9 above)
Comment (wowtor) [reply lir/algorithms/bootstraps.py]: Of eventueel een aparte PR naar deze branch. Twee vrijwel identieke klassen naar main mergen is ook twee keer hetzelfde reviewen en zolang die tweede PR er niet is, is het code dubbel onderhouden.

Comment (PimMeulensteen) [reply lir/algorithms/bootstraps.py]: Ik heb het nu gewoon als commit bij de PR gedaan.

File: lir/algorithms/bootstraps.py [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+from typing import Any, Self
+
+import numpy as np
+from scipy.interpolate import interp1d
+
+from lir.data.models import FeatureData, LLRData
+from lir.transform.pipeline import Pipeline
+
+
+class Bootstrap(Pipeline):
```
Comment (wowtor) [inline lir/algorithms/bootstraps.py]: abstract class

```suggestion
class Bootstrap(Pipeline, ABC):
```

File: lir/algorithms/bootstraps.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        self.f_delta_interval_upper = None
+
+    @abstractmethod
+    def get_bootstrap_data(self, instances: FeatureData) -> FeatureData:
+        """Get the data points to use for interval estimation.
+
+        param instances: FeatureData: The feature data to fit the bootstrap system on.
+        return FeatureData: The feature data to use for interval estimation.
+        """
+        pass
```
Comment (wowtor) [inline lir/algorithms/bootstraps.py]: Per conventie (zie https://docs.python.org/3/library/exceptions.html#NotImplementedError)

```suggestion
        raise NotImplementedError
```

File: lir/algorithms/bootstraps.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+        lower = intervals[0] - x_vals
+        upper = intervals[1] - x_vals
+
+        # Use the numeric 1D feature array as x for interpolation (not the FeatureData object).
+        self.f_delta_interval_lower = interp1d(x_vals, lower, bounds_error=False, fill_value=(lower[0], lower[-1]))
+        self.f_delta_interval_upper = interp1d(x_vals, upper, bounds_error=False, fill_value=(upper[0], upper[-1]))
+        return self
+
+
+class BootstrapAtData(Bootstrap):
```
Comment (wowtor) [inline lir/algorithms/bootstraps.py:137]: Veel overzichtelijker zo. Nu is ook meteen duidelijk waar het verschil zit tussen de twee bootstrap implementaties!

**Review by wowtor — APPROVED** [review]
Mooi!
